### PR TITLE
Subdomain basic statistics . 

### LINKED ISSUE #1923: [TASK] Basicstatistic subdomain
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
It would be extremely useful to allow the basic statistics to accept a "subdomain" definition (a grid in the input space where to perform the computation of the moments and correlation, sensitivity matrices).

For example, 
allowing for the definition of a grid (like in the grid sampler)

**Describe the solution you'd like**
Possibility to compute "sensitivity coefficients" (e.g. reactivity coefficients) based on a Grid sampling

e.g.

T (100,200,300,400,500,600)
dk/dT (200,300,400,500)

**Describe alternatives you've considered**
To split the dataset by hand and apply a basic statistic to each of the separate datasets 

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by aalfonsi**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1923 

##### What are the significant changes in functionality due to this change request?
Added a new post processor to compute subdomain basic statistics leveraging RAVEN basic statistics. 

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: ravenframework/Models/PostProcessors/BasicStatistics.py [hunk 1]
```diff
@@ -1551,7 +1562,8 @@ def run(self, inputIn):
       @ Out, outputSet, xarray.Dataset or dictionary, dataset or dictionary containing the results
     """
     inputData = self.inputToInternal(inputIn)
-    outputSet = self.__runLocal(inputData)
+    outputSet = self._runLocal(inputData)
+    print(outputSet)
```
Comment (aalfonsi) [inline ravenframework/Models/PostProcessors/BasicStatistics.py:1566]:
```suggestion
```

**Review by mandd — CHANGES_REQUESTED** [review]

**Comment by mandd** [comment]
I have left my initial set of comments. One of my main questions is why a new PP class was added rather than extending the directly the basic statistics?

File: doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      sen_x01_a, sen_x01_b,
+      sen_x02_a, sen_x02_b
+    </Output>
+  </PointSet>
+</Simulation>
+\end{lstlisting}
+
+In this case, the RAVEN variables ``mean\_x01, mean\_x02, sen\_x01\_a, sen\_x02\_a, sen\_x01\_b, sen\_x02\_b''
+will be computed for each cell defined above (2 cells in this case).
+The results will be associated to the mid-point of each point. For the example above:
```
Comment (aalfonsi) [inline doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:72]:
```suggestion
The results will be associated to the mid-point of each cell. For the example above:
```

File: doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+\begin{itemize}
```
[the remark this answers is not in the document]
Comment (aalfonsi) [reply doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:73]: The interval of the fist grid is
[10,100] with 2 steps:
`|---------|---------|`
`10        55       100`
midpoints:
 `------|--------|------`
 `      32.5     77.5 `

File: doc/user_manual/raven_user_manual.tex [hunk 4]
```diff
@@ -90,6 +90,19 @@
...
+\setlist[itemize,1]{label=$\bullet$}
+\setlist[itemize,2]{label=$\bullet$}
+\setlist[itemize,3]{label=$\bullet$}
+\setlist[itemize,4]{label=$\bullet$}
+\setlist[itemize,5]{label=$\bullet$}
+\setlist[itemize,6]{label=$\bullet$}
+\setlist[itemize,7]{label=$\bullet$}
+\setlist[itemize,8]{label=$\bullet$}
+\renewlist{itemize}{itemize}{8}
+
```
[the remark this answers is not in the document]
Comment (aalfonsi) [reply doc/user_manual/raven_user_manual.tex:105]: I added a comment at the begin of the command:
```the following allows for 8 levels list nesting```

This allows to add an additional level of nesting for the "itemize" command

File: doc/user_manual/sampler.tex [hunk 5]
```diff
@@ -118,6 +118,34 @@ \section{Samplers}
...
+      \\ $stepSize=(upperBound - lowerBound)/steps$
+  \item \xmlAttr{construction}\textbf{\texttt{=}}\xmlString{custom}.
+    The grid will be directly specified by the user.
+    No additional attributes are needed.
+    This construction type requires that the \xmlNode{grid} node contains
+    the actual mesh bins.
+    For example, if the grid \xmlAttr{type} is \xmlString{value}, in the body
+    of \xmlNode{grid}, the user will specify the values representing the nodes of the grid.
+\end{itemize}
+}
```
[the remark this answers is not in the document]
Comment (aalfonsi) [reply doc/user_manual/sampler.tex:148]: yeah probably not the right location. Moved to postprocessors.tex

File: doc/user_manual/sampler.tex [hunk 6]
```diff
@@ -234,6 +262,21 @@ \section{Samplers}
   information is requested.}
 }
 
+\newcommand{\gridDescriptionNoCFD}
+{
```
[the remark this answers is not in the document]
Comment (aalfonsi) [reply doc/user_manual/sampler.tex:266]: Typo. CFD -> CDF (Cumulative Dist Function)

File: doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex (hunk 2 above)
[the remark this answers is not in the document]
Comment (aalfonsi) [reply doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex]: fixed

**Comment by aalfonsi** [comment]
> I have left my initial set of comments. One of my main questions is why a new PP class was added rather than extending the directly the basic statistics?

@mandd addressed your comments. The reason is mostly because it would have required an additional refactoring of the basic statistics and it is already quite complex. The advantage of having this PP is that the computation is encapsulated in the BasicStatistics and, in the future, the calculation for each "cell" can be parallelized if needed.

File: doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex (hunk 3 above)
[the remark this answers is not in the document]
Comment (mandd) [reply doc/user_manual/PostProcessors/SubdomainBasicStatistics.tex:73]: gotcha

**Review by mandd — APPROVED** [review]

**Comment by wangcj05** [comment]
Hi @aalfonsi: Is it possible that the subdomain information (the grid lower and upper bound) also get stored in the DataObjects, for example in the metadata? I would recommend to preserve this information in the output DataObject.

**Comment by aalfonsi** [comment]
> Hi @aalfonsi: Is it possible that the subdomain information (the grid lower and upper bound) also get stored in the DataObjects, for example in the metadata? I would recommend to preserve this information in the output DataObject.

@wangcj05 can you open an issue and I can address it in a followup MR?

**Pull request merged**